server/list.js

```diff
@@ -102,48 +102,57 @@ function getOptions(id) {
   }
 }
 
-async function fetchAllFiles({nextPageToken: pageToken, listSoFar = [], parentIds = [driveId], driveType = 'team', drive} = {}) {
-  const options = getOptions(parentIds)
+async function fetchAllFiles({nextPageToken: pageToken, nextListSoFar: listSoFar = [], nextParentId: parentId = driveId, nextParentName: parentName = 'root', driveType = 'team', drive} = {}) {
+  const options = getOptions(parentId)
 
   if (pageToken) {
     options.pageToken = pageToken
   }
 
-  log.debug(`searching for files > ${listSoFar.length}`)
+  log.debug(`searching for files > ${listSoFar.length} in "${parentName}"`)
 
   // Gets files in single folder (shared) or files listed in single page of response (team)
   const {data} = await drive.files.list(options)
 
   const {files, nextPageToken} = data
-  const combined = listSoFar.concat(files)
+  const sublist = []
 
-  // If there is more data the API has not returned for the query, the request needs to continue
-  if (nextPageToken) {
-    return fetchAllFiles({
-      nextPageToken,
-      listSoFar: combined,
-      drive
-    })
+  if (driveType === 'folder') {
+    // Continue searching if shared folder, since API only returns contents of the immediate parent folder
+    // Find folders that have not yet been searched
+    const folders = files.filter((item) =>
+      item.mimeType === 'application/vnd.google-apps.folder' && parentId === item.parents[0])
+
+    if (folders.length > 0) {
+      // Avoid Promise.all to avoid hitting google API rate limits
+      await folders.reduce((acc, folder) => {
+        return acc.then((files) => fetchAllFiles({
+          drive,
+          driveType: 'folder',
+          sublist,
+          nextPageToken,
+          nextParentId: folder.id,
+          nextParentName: folder.name
+        }).then((newFiles) => files.concat(newFiles)))
+      }, Promise.resolve([]))
```

Comment:
a separate issue I ran into (after fixing the underlying issue) is that for large document trees, eventually the list of parentIds would get so long that the API call would fail. so this breaks them out into individual calls to avoid that error case